H3Hexagon: add support for coverage.

1chandu opened the pull request:
<!-- For feature, feature enhancement or bug fix, create an issue first and finish To Do List there -->
<!-- Anything doesn't work as expected is a bug, including code, doc and test -->
For #3209 

<!-- For other PRs without open issue -->
#### Background
- Forward `coverage` prop to `ColumnLayer`
- Scale polygon vertices when using `PolygonLayer`.

<!-- For all the PRs -->
#### Change List
- H3Hexagon: add support for coverage.


![H3Coverage](https://user-images.githubusercontent.com/9502731/59066629-57a44000-8864-11e9-889a-982c5be4b78b.gif)

Pessimistress commented on modules/geo-layers/src/h3-layers/h3-hexagon-layer.js:
```diff
@@ -22,9 +23,27 @@ function getHexagonCentroid(getHexagon, object, objectInfo) {
+  const [lat, lng] = h3ToGeo(hexId);
+  // get distinct vertex objects,
+  // `h3ToGeoBoundary` returns same array object for first and last vertex (closed polygon)
+  const uniqueVertices = [...new Set(vertices)];
```
- Don't use spread operator in this repo
- This is unnecessary. the construction of new objects is more expensive than transforming one additional point. `h3ToPolygon` should consistently return the same format regardless of `coverage`.

1chandu replied:
`h3ToGeoBoundary` is returning same object (Array object) for first and last vertex , changing one changes the other, I will have WA here (just skip the last), and will open bug for h3-js.

Pessimistress replied:
I see. I don't think this is necessarily a bug for h3-js, but indeed a catch.

Pessimistress commented on modules/geo-layers/src/h3-layers/h3-hexagon-layer.js:
```diff
@@ -22,9 +23,27 @@ function getHexagonCentroid(getHexagon, object, objectInfo) {
+  // `h3ToGeoBoundary` returns same array object for first and last vertex (closed polygon)
+  const uniqueVertices = [...new Set(vertices)];
+  for (const vertex of uniqueVertices) {
+    const scaledPt = lerp([lng, lat], vertex, factor);
```
Do not create new array in for loop. You can simply call `lerp` twice.

Pessimistress commented on modules/geo-layers/src/h3-layers/h3-hexagon-layer.js:
```diff
@@ -36,6 +55,18 @@ function h3ToPolygon(hexId) {
+  if (!currentTrigger) {
+    trigger = newTrigger;
+  } else if (Array.isArray(currentTrigger)) {
+    trigger = [...currentTrigger, newTrigger];
```
Perf: use `concat`

Pessimistress commented on modules/geo-layers/src/h3-layers/h3-hexagon-layer.js:
```diff
@@ -36,6 +55,18 @@ function h3ToPolygon(hexId) {
+  } else if (Array.isArray(currentTrigger)) {
+    trigger = [...currentTrigger, newTrigger];
+  } else {
+    trigger = [currentTrigger, newTrigger];
```
`currentTrigger` can be an object.

1chandu commented:
@Pessimistress , addressed comments.

deck.gl 7.0 supported the coverage for non `highPrecision` mode (we were forwarding the prop), any concerns cherry-picking this into 7.1-release branch ?

Pessimistress commented on modules/geo-layers/src/h3-layers/h3-hexagon-layer.js:
```diff
@@ -22,9 +23,26 @@ function getHexagonCentroid(getHexagon, object, objectInfo) {
+  const [lat, lng] = h3ToGeo(hexId);
+  // `h3ToGeoBoundary` returns same array object for first and last vertex (closed polygon),
+  // hence skip scaling the last vertex
+  log.assert(vertices[0] === vertices[vertices.length - 1]);
```
Perf: save `vertices.length` in a variable:

```
const n = vertices.length;
for (let i = 1; i < n; i++) {
  ...
}
```

Pessimistress commented on modules/geo-layers/src/h3-layers/h3-hexagon-layer.js:
```diff
@@ -36,6 +54,20 @@ function h3ToPolygon(hexId) {
   return vertices;
 }
 
+function mergeTriggers(currentTrigger, newTrigger) {
```
What's the point of creating a `newTrigger` object if you're just going to use `newTrigger.cover`?

1chandu replied:
Either I have to pass value and do {coverage: value} or pass an object and access it as object.coverage. Changed it to latter former case.

Pessimistress commented on modules/geo-layers/src/h3-layers/h3-hexagon-layer.js:
```diff
@@ -36,6 +54,20 @@ function h3ToPolygon(hexId) {
+  if (currentTrigger === undefined || currentTrigger === null) {
+    trigger = newTrigger.coverage;
+  } else if (Array.isArray(currentTrigger)) {
+    trigger = [].concat(currentTrigger, newTrigger.coverage);
```
`Object.assign({}, <array>)` works, you don't have to handle it as a special case.

Pessimistress commented on modules/geo-layers/src/h3-layers/h3-hexagon-layer.js:
```diff
@@ -36,6 +54,20 @@ function h3ToPolygon(hexId) {
+  } else if (typeof currentTrigger === 'object') {
+    trigger = Object.assign({}, currentTrigger, newTrigger);
+  } else {
+    trigger = [currentTrigger, newTrigger.coverage];
```
`{currentTrigger, coverage}`

Pessimistress commented on modules/geo-layers/src/h3-layers/h3-hexagon-layer.js:
```diff
@@ -22,9 +23,26 @@ function getHexagonCentroid(getHexagon, object, objectInfo) {
 }
 
-function h3ToPolygon(hexId) {
+function scalePolygon(hexId, vertices, factor) {
```
Add a test for this function?

Pessimistress commented on modules/geo-layers/src/h3-layers/h3-hexagon-layer.js:
```diff
@@ -22,9 +23,27 @@ function getHexagonCentroid(getHexagon, object, objectInfo) {
+
+  if (coverage !== 1) {
+    // updates array object elements of vertices array.
+    scalePolygon(hexId, vertices, coverage);
```
This breaks if the hexagon crosses the 180th meridian. You need to set the refLng to that of the center, normalize it before scaling.

1chandu replied:
Done.

Pessimistress commented on modules/geo-layers/src/h3-layers/h3-hexagon-layer.js:
```diff
@@ -22,9 +23,27 @@ function getHexagonCentroid(getHexagon, object, objectInfo) {
+export function scalePolygon(hexId, vertices, factor) {
+  const [lat, lng] = h3ToGeo(hexId);
+  const vertexCount = vertices.length;
+  log.assert(vertices[0] === vertices[vertexCount - 1]);
```
This assert will break the layer of h3-js makes a change in the future. It won't be the user's fault and it won't provide any useful error message. Can we just adapt the for loop based on whether these two points are the same object?

1chandu replied:
Done.

Pessimistress commented on test/modules/geo-layers/h3-layers.spec.js:
```diff
@@ -50,6 +54,121 @@ test('H3HexagonLayer', t => {
   t.end();
 });
 
+test.only('H3HexagonLayer#mergeTriggers', t => {
```
test.only

1chandu commented on modules/geo-layers/src/h3-layers/h3-hexagon-layer.js:
```diff
@@ -33,9 +28,57 @@ function h3ToPolygon(hexId) {
+  }
+
+  // normalize with respect start vertex
+  normalizeLongitudes(vertices);
```
I wonder we should move this into `PolygonLayer` or it's sublayers.
